locator/topology.cc

```diff
@@ -72,14 +73,12 @@ topology::topology(config cfg)
 {
     tlogger.trace("topology[{}]: constructing using config: host_id={} endpoint={} dc={} rack={}", fmt::ptr(this),
             cfg.this_host_id, cfg.this_endpoint, cfg.local_dc_rack.dc, cfg.local_dc_rack.rack);
-    if (cfg.this_host_id || cfg.this_endpoint != inet_address{}) {
-        add_node(node::make(this, cfg.this_host_id, cfg.this_endpoint, cfg.local_dc_rack, node::state::joining, node::this_node::yes));
-    }
```

Comment:
But then aren't we back to the problem of losing `this_node` after clear_gently()?

I think that even before this patch, replacing with the same ip will put `this_node` on the bootstrapping node into leaving state, because we lookup by IP when changing state. There is no way to hold two nodes with the same IP currently because data structures are indexed by IP.

The only cure for nervousness is to identify nodes by host_id and have the state of topology not depend on the order of operations. If we identify by host id, and we want to keep both nodes, we should put the local host id into topology::config before populating it for the topology to distinguish the nodes. But that requires also changing the interface for updates to accept host id instead of IP.

Reply:
That's the direction we're aiming at but we're not quite there yet.

We need to test the hello out of this change to make sure it doesn't introduce regressions.
I recommend the following dtest suites: repair_based_node_operations.py replace_address_test.py update_cluster_layout_tests.py

Reply:
I ran repair_based_node_operations.py and replace_address_test.py and they passed except for replace_address_test.py::TestReplaceAddress::test_replace_node_no_hibernate_state, which is a preexisting issue (reproduces without the PR): https://github.com/scylladb/scylladb/issues/13543